src/app/public/modules/popover/popover.directive.spec.ts

```diff
@@ -1,309 +1,850 @@
-import {
-  DebugElement
-} from '@angular/core';
-
 import {
   ComponentFixture,
-  TestBed,
-  tick,
   fakeAsync,
-  async
+  inject,
+  TestBed,
+  tick
 } from '@angular/core/testing';
 
-import {
-  By
-} from '@angular/platform-browser';
-
-import {
-  NoopAnimationsModule
-} from '@angular/platform-browser/animations';
-
 import {
   expect,
   SkyAppTestUtility
 } from '@skyux-sdk/testing';
 
 import {
-  SkyWindowRefService
+  SkyAffixService,
+  SkyOverlayService
 } from '@skyux/core';
 
 import {
-  SkyPopoverTestComponent
+  Subject
+} from 'rxjs/Subject';
+
+import {
+  PopoverFixtureComponent
 } from './fixtures/popover.component.fixture';
 
 import {
-  SkyPopoverModule
-} from './popover.module';
+  PopoverFixturesModule
+} from './fixtures/popover.module.fixture';
 
 import {
   SkyPopoverMessageType
 } from './types/popover-message-type';
 
-import {
-  SkyPopoverDirective
-} from './popover.directive';
-
 import {
   SkyPopoverAdapterService
 } from './popover-adapter.service';
 
-class MockWindowService {
-  public getWindow(): any {
-    return {
-      setTimeout(callback: Function) {
-        callback();
-      }
-    };
-  }
-}
-
-describe('SkyPopoverDirective', () => {
-  let fixture: ComponentFixture<SkyPopoverTestComponent>;
-  let directiveElements: DebugElement[];
-  let mockWindowService: MockWindowService;
-
-  function validateTriggerOpensPopover(
-    elIndex: number,
-    openTrigger: string,
-    closeTrigger: string
-  ) {
-    const caller = directiveElements[elIndex];
-    const callerInstance = caller.injector.get(SkyPopoverDirective);
-
-    const positionNextToSpy = spyOn(callerInstance.skyPopover, 'positionNextTo');
-    const closeSpy = spyOn(callerInstance.skyPopover, 'close');
-
-    // The popover should only execute hover events if it is set to 'mouseenter'.
-    if (openTrigger !== 'mouseenter') {
-      SkyAppTestUtility.fireDomEvent(caller.nativeElement, 'mouseenter');
-      expect(positionNextToSpy).not.toHaveBeenCalled();
-    }
+describe('Popover directive', () => {
 
-    SkyAppTestUtility.fireDomEvent(caller.nativeElement, openTrigger);
-    expect(positionNextToSpy).toHaveBeenCalled();
+  let fixture: ComponentFixture<PopoverFixtureComponent>;
 
-    callerInstance.skyPopover.isOpen = true;
+  function getCallerElement(): HTMLButtonElement {
```

Comment:
Would you mind wrapping these helpers in a collapsable region?

Reply:
You got it.